docs/model/index.md

```diff
@@ -0,0 +1,36 @@
+# farmOS Data Model
+
+## Goals
+
+The goal of the farmOS data model is to represent and organize a wide variety
+of agricultural and ecological systems to allow for easy access and analysis.
+This ranges from small-scale garden records to industrial agriculture to
+national forestry management.
+
+farmOS can model data that was recorded via manual data entry, as well as data
+streams from sensors or other applications.
+
+## Data types
+
+Data is organized into a set of high-level data types. The two primary record
+keeping data types are **Assets** and **Logs**. Other types include
+**Quantities**, **Terms**, **Plans**, and **Users**.
+
+- [Assets](/model/type/asset)
+- [Logs](/model/type/log)
+- [Quantities](/model/type/quantity)
+- [Terms](/model/type/term)
+- [Plans](/model/type/plan)
+- [Users](/model/type/user)
+
+## Logic
```

Comment:
Ok, I know I've caused a lot of fuss over "logic" in the past, but what if we rephrased this as "Operations" instead of "Logic"? So hear me out...

First, logic seems a little too broad to me in these examples. There is a _lot_ of logic in farmOS, but I feel like we're trying to encapsulate something much more specific here, with regards to the actual _data structures_. For reference, here's how Wikipedia defines a [Data structure](https://en.wikipedia.org/wiki/Data_structure):

> More precisely, a data structure is a collection of data values, the relationships among them, and the functions or __operations__ that can be applied to the data. [emphasis added]

A good example here, outside of farmOS, would be to think about the built-in data structures in a language, like numbers, strings and arrays. There are different operators for different data structures, for example:

```php
// Numeric operators
4 + 2;
4 - 2;

// String operator
"Hello" . "World";

// Array operator
array("a" => 42, "b" => "foo") + array("c" => "bar");
```

And these can get proportionally more complex as the data structures get more complex.

For farmOS, we could think of a _movement_ as a specific type of operation, where the operands are an asset and a log, with the result being a given change in the asset's location. Same would go for _group membership_.

I guess this is splitting hairs a bit, and it's not to say there still isn't logic involved in each operation, but I feel like it could help to frame the different types of operations, based on the data structures they operate on. This could eventually help inform future work on inventory, as specific operations on assets and quantities.

And despite the fact that "operation" can have a rather mathematical connotation to it, it has a second connotation to the running of a business that might make it more intuitive to non-programmers.

Reply:
This is an interesting way to look at it.

> I feel like we're trying to encapsulate something much more specific here, with regards to the actual data structures.

Yes!

> There are different operators for different data structures

True - but this may be more atomic than what we're going for with this documentation section.

It isn't so much about describing what kind of operations can be performed (in some ways that's what the "Logs" section covers). Instead, it tries to describe "how the granular records themselves, when looked at together, bubble up greater meaning".

> For farmOS, we could think of a movement as a specific type of operation, where the operands are an asset and a log, with the result being a given change in the asset's location. Same would go for group membership.

Maybe these are two ways to say the same thing, but I feel like there is a key distinction here...

I think about farmOS as similar to a ledger. It has accounts (assets) and transactions (logs). The "balance" of the account is calculated based on the transactions, and this "balance" can be determined for any point in time. In farmOS, the "balance" is whatever "pseudo property" it is we're trying to determine, like location, group membership, inventory, etc.

"Moving" an animal can be considered an "operation" - but the act of performing that is accomplished by creating a log. And so the log itself *is* the operation. However, *how* the location is determined based on logs is the logic.

In order for logs and assets to have any meaning, we have to define how a set of logs can be looked at collectively to determine an asset's "state" at *any point in time*. Perhaps we don't emphasize that point enough in the current logic docs. The logic can be used to get an asset's "current" location, but it is the same exact logic used to get the asset's location at a given timestamp in the past or future.

So, I see "logic" as meaning "we've defined how to compute a given asset property at a given timestamp, and this is the logic that is followed to do that".

In some ways, this logic itself is a form of convention - one that has been baked into farmOS code. But ultimately, it is a definition for how to pull meaning out of the lower level records.

The main reason to document this logic is so that *other* implementations of the farmOS data model can replicate the same logic if they need to. If someone were to pull all their farmOS records down to JSON files in a directory, and wanted to compute the location of an asset from them, they would need to perform the same logic.

> it could help to frame the different types of operations, based on the data structures they operate on

Also worth noting that a single log can perform multiple "operations". You can have a log that assigns group membership to an asset, changes its location, and increments its inventory all at once. These are just three separate "balance calculations" that use the same log in their equations.

>  it has a second connotation to the running of a business

I suppose the other hesitation I have is that "operations" is used to mean something more domain specific in farming/agriculture, and is perhaps more analogous to what logs themselves represent (farming operations, events, things that were done on the farm, etc). I worry that a farmer coming to these docs and seeing "Operations" would expect to find information about the type of operations they can record in farmOS, which would actually be found under "log types".

---

All that said, I agree "logic" might be a bit opaque. What is it they say about the 2 hardest problems in programming? 1. cache invalidation, 2. **naming things**, and 3. off-by-one errors. :-)

Reply:
Taking a step back, I admit that "operation" and "logic" are kind of interchangeable in my usage. But the main thrust of my argument might boil down to: "operation" implies logic that is applied to a particular set of data structures, whereas "logic" alone does not imply the associated data structures, or at least does not have as strong an implication.

> I think about farmOS as similar to a ledger. It has accounts (assets) and transactions (logs).

Right, but I think accounts and transactions are still plain data structures by themselves, which require an operation to determine the new balance (another data structure).

> And so the log itself is the operation.

I don't think it is though. A log can be either a data structure or an operation (again, "operation" here implying "logic that is applied to a particular set of data structures"), not both. I think we've come down pretty squarely on the side that a log is a data structure, insofar as we can send them over the wire as JSON. Perhaps it would help to clarify that operations can have multiple operands, so the movement operation has two operands: an asset to be moved, and a movement log.

> but the act of performing that is accomplished by creating a log

We might be saying the same thing here, but I'm trying to say that the "operation" is the verb in this sentence, whereas the  "log" is the noun (the "asset" is a noun, too, though unstated here). I don't know if that's helpful, so I won't get hung up on that analogy if it doesn't resonate with you.

> The main reason to document this logic is so that other implementations of the farmOS data model can replicate the same logic if they need to. If someone were to pull all their farmOS records down to JSON files in a directory, and wanted to compute the location of an asset from them, they would need to perform the same logic.

I think I'm fully in agreement with you on all of this, except I would just substitute the words "operation" for "logic". Just arguing that it's a more precise usage, which clarifies the meaning.

> Also worth noting that a single log can perform multiple "operations".

Again, I think we have to distinguish that the log is data; it doesn't _do_, it _is_. The "operation" is what's performed on the log. And multiple operations can be performed on the same data. I can _add_ 5 and 6 just as easily as I can _subtract_ 5 and 6.

Maybe it would be helpful to look at it this way. I'm thinking of the operation as the middle term in both of these expressions:

```
5 + 6
log MOVES asset
```

Where both `+` and `MOVES` are merely symbols representing a computational operation. Does that make sense?

So what I'm arguing is that "operation" is a better word for the computation performed by such an infix operator. We could call that logic, too, but the difference, in my mind, is that operation is more specific to the data structures it can be performed on. In other words, it implies that the `MOVES` operation above is only used with a log (on the left) and an asset (on the right). Both of these expressions would then be considered illegal:

```
quantity MOVES asset
asset MOVES log
```

Those expressions don't make sense, because `MOVES` is an operation specific to the data structures it operates on.

I think the biggest sticking point here is whether "logic" has the same implications. In my mind, it doesn't, but that could be a subjective thing. I'd be curious to know what other people think on this.

Reply:
> We might be saying the same thing here, but I'm trying to say that the "operation" is the verb in this sentence, whereas the "log" is the noun (the "asset" is a noun, too, though unstated here).

Yea I think we're in sync. :-)

"Operation" might just be too loaded of a term. I do appreciate how it tries to relate these ideas to mathematical operations though - I certainly agree there. But does it confuse more than it helps, I wonder?

The goal of having a dedicated "logic/operations" section, and not just adding info to the "record types" themselves, is to explicitly highlight that this "logic/operations" is a separate layer. I think both terms serve that - but I worry that "operations" might be initially confusing.

I'd love to hear what others think of these ideas!

---

> `log MOVES asset`

I like the simplicity of describing things this way - but I'm curious how this can also capture the deeper parameters involved. Specifically, a log only `MOVES` an asset "if it has a status of 'done' and has a timestamp less than or equal to the reference point (which by default is the current timestamp)". It's also possible to ask "where was an asset in the past" and "where will an asset be in the future" (which includes "pending" logs).

I like the idea of describing some of these things as simple "operations" but I think the "logic" *includes* operations, but also includes decisions about how those operations will be interpreted, if that makes sense?

In my mind the "farmOS data model logic" defines not only the allowable operations, but also how we are deciding to interpret the results of these equations. So I think there is a level of meaning that goes beyond just the mathematical analogy? But maybe I'm overthinking it. :-)

---

Perhaps another point I'd like to highlight, though, is that the "logic" section does have some other stuff in it that doesn't fit neatly into the "operations" idea. For example, at the bottom of the Location logic page:

> When a Log is saved without a geometry, and it references locations that have geometries, the locations geometries will be copied to the Log.

Now, maybe that piece in particular is out of place there... but I think little things like that should also be documented, and the "logic" section seemed like the right place for it. But perhaps that's a separate thread. :-)

Reply:
> But does it confuse more than it helps, I wonder?

Yea, I think that's what really remains to be seen.

> I like the simplicity of describing things this way - but I'm curious how this can also capture the deeper parameters involved. Specifically, a log only `MOVES` an asset "if it has a status of 'done' and has a timestamp less than or equal to the reference point (which by default is the current timestamp)". It's also possible to ask "where was an asset in the past" and "where will an asset be in the future" (which includes "pending" logs).

I think part of what I like about this framing is it forces us to be explicit about the deeper parameters involved. If, instead of just one log we're talking about a collection of logs that must all be taken together to get the current location of an asset, or its location at a given point in time, we have to say so. And we have to distinguish that operation from the operation that only requires one log. Also, we could express the operation on many logs in terms of the operation on one log, which gives a level of composition.

> but I think the "logic" includes operations

I mean, I could argue that operations includes logic (and I am). I also think operations can be composed of other operations. But I'm going to reserve my judgment on that for the moment.

> In my mind the "farmOS data model logic" defines not only the allowable operations, but also how we are deciding to interpret the results of these equations.

I don't think this rules out the idea of operations. I didn't address it in my previous example, but operations have results. And there are lots of ways that can happen, which is why I avoided it, b/c I didn't want to get bogged down in that. But you can have something like an expression which evaluates to the result:

```lisp
(+ 4 5) ;; evaluates to 9
(MOVE log asset) ;; evaluates to an asset w/ new geometry
```

You can have an assignment type of operation which mutates the original variable holding that value:

```go
$x := 4;
$x += 5; // $x is now 9
$asset := create_asset();
$asset.move($log); // mutates the asset's geometry 
```

Anyways, I think those are implementation details, but I still think the core idea of operations requires that those results be explicit in a way that logic doesn't imply. Even if the result is that nothing changed:

```js
x += 0;
```

At least, I struggle to think of a concept of "operation" that does _not_ imply a result, but perhaps that's just me.

> Now, maybe that piece in particular is out of place there... but I think little things like that should also be documented, and the "logic" section seemed like the right place for it.

I agree! And again, I think conceptualizing these as operations only reinforces that requirement. If the documentation fails to specify all the results of an operation on its operands, that's a failure of the documentation, and I think becomes more clear when we think of these as operations instead of logic.